### PR TITLE
[wasm] Support for delivering a javascript payload

### PR DESCRIPTION
**Pull request by kjpou1**
Functionality:

- Facilitates developer add-on integration within the wasm runtime.
- Reduces the amount of assets that need to be delivered and included in client html pages specifically _<script>_ tags.

Modifications:

- Add support to the `driver.c` module that will load a javascript payload from the assembly file that is loaded.
- Add functionality to `mono_wasm_assembly_load` that searches for a method called `WebAssembly.Library.ScriptsSupport:Initialize` which once invoked will return the javascript payload to be loaded into the current environment.

```
namespace WebAssembly.Library
{
    public static class ScriptsSupport
    {

        // Load our support library from the embedded resources.
        public static string Initialize()
        {
// Load payload here
        }
    }
}

```

- The payload can be delivered as an `Embedded Resource`.

```

            var assembly = Assembly.GetAssembly(typeof(Sample.Program));
            var resourceNames = assembly.GetManifestResourceNames();
            foreach (var res in resourceNames)
            {
                using (Stream stream = assembly.GetManifestResourceStream(res))
                using (StreamReader reader = new StreamReader(stream))
                {
                    string result = reader.ReadToEnd();

                    if (res.EndsWith("library_support.js"))
                    {
                        return result;
                    }
                }
            }
            return null;

```

- example `library_support.js` Embedded Resource:

```

(function(global) {
    
    GreeterLibrary = function () {
        return new GreeterLibrary.init();
    }

    GreeterLibrary.prototype = {
        sayHello: function(greeting) {
            console.log(greeting);
        },

    };

    GreeterLibrary.init = function ()
    {
        var self = this;
        // place other initialization here.
        console.log("initialized GreeterLibrary");
    }

    GreeterLibrary.init.prototype = GreeterLibrary.prototype;

    // make the library available to the outside world if needed.
    global.GreeterLibrary = GreeterLibrary;

    
}(window));

// Initialize our GreeterLibary
window.Greeter = new GreeterLibrary();

``` 

- When payload is returned back to the driver program it will then be loaded into the environment using EMSCRIPTEN's `EM_ASM_INT` functionality.

- To reference the `Greeter` object using the wasm sdk `WebAssembly.Bindings` (optional) you can do the following:

```
            var Greeter = (JSObject)Runtime.GetGlobalObject("Greeter");
            Greeter.Invoke("sayHello", "Hello World");
```

### CODE REVIEW

File: sdks/wasm/driver.c [hunk 1]
```diff
@@ -143,6 +143,32 @@ mono_wasm_setenv (const char *name, const char *value)
...
+	const char *aname = mono_assembly_name_get_name (assembly_name);
+
+	// Can not reflect over mscorlib
+	if (strcmp ("mscorlib", aname))  // not equal
+	{
+		MonoClass* klass = mono_class_from_name(mono_assembly_get_image(assembly), "WebAssembly.Library", "ScriptsSupport");
+		if (klass)
+		{
+			MonoMethod* init_method = mono_class_get_method_from_name(klass, "Initialize", -1);
+			MonoException* exc = NULL;
```
[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/driver.c]: How should we get around that?  Can we use [Preserve] or something?

[the remark this answers is not in the document]
Comment (vargaz) [reply sdks/wasm/driver.c]: Another thing is, assembly_load_hooks () are called from the runtime from deep within the code, its generally not a good idea to invoke managed code from them.

[the remark this answers is not in the document]
Comment (vargaz) [reply sdks/wasm/driver.c]: What are you trying to achieve here ?

[the remark this answers is not in the document]
Comment (kjpou1) [reply sdks/wasm/driver.c]: Let me take another stab at it then.

**Comment by kg** [comment]
You should consider not using eval() for this purpose, even if we already use it extensively. Content Security Policies can disable eval for security reasons, which would make it impossible to replace <script> payloads with this feature.

Programmatically creating and inserting a <script> element is ideal for large blocks of code because that integrates best with developer tools cross-browser (you can also set a url for it using a special comment). For smaller blocks of code I recommend using new Function. The CSP rule for inline <script> and for eval-esque APIs are separate. (I think new Function is under the same permission as eval.)

Also FYI, some ways of loading code are asynchronous in Chrome even though they appear not to be. Keep that in mind if you do adopt the use of methods other than eval.

**Pull request closed without merging**